### DISCUSSION TITLE
cpp lexer adjustments

### DISCUSSION THREAD

**Discussion by snoopy** [orbitalquark/scintillua#111] [Q&A] [answered]
I'm trying to make some adjustments to the cpp lexer and ran into some problems.

I'd like to remove the highlight for `std::` from STL types. Highlighting that part only increases clutter and takes away from the info you are looking for IMO.
Removing the `std::` part from the lexer works but will also highlight unrelated tokens of the same name. Is there a way to match the full token but only highlight the part without `std::`? E.g. match `std::vector` but only highlight `vector`.

Related to this question, I also want to highlight custom types at the end of a scope chain. E.g. in `foo::bar::type`, `type` should be highlighted. This can be achieved with a rule like the following:
```lua
lex:add_rule('custom_type', lex:tag(lexer.TYPE .. '.custom', '::' * lexer.word * -S('(:')))
```
The problem is that this also matches the leading `::` part which it should not. Any way to avoid this? I've tried `#('::')` and `#P('::')`, both of which produce wrong results.

---

Some other questions that might make sense for the default lexer as well.

* How to highlight angle braces?
    * In C++ it makes sense to highlight `<` and `>` but it doesn't work out of the box. I've tried adding `lexer.property['scintillua.angle.braces'] = '1'` which will do the trick but also adds autopairing which is not wanted here.
* Should template functions be highlighted as such?
    * E.g. things like `std::make_shared<int>()` are currently not highlighted as functions

Lastly, are there still plans to use treesitter over or maybe alongside lua lexers? @orbitalquark

**Comment by orbitalquark** [comment, marked as answer]
To match 'std::' but not highlight it, you could do something like this:

```
lex:tag(lexer.DEFAULT, 'std::') * lex:tag(lexer.TYPE .. '.stl', lex:word_match(lexer.TYPE .. '.stl'))
```

For the related question, maybe

```
lex:add_rule('custom_type', lex:tag(lexer.DEFAULT, '::') * lex:tag(lexer.TYPE .. '.custom', lexer.word * -S('(:')))
```

? I'm not sure I understand exactly what you're trying to do.

You can highlight angle braces using `lexer.property['scintillua.angle.braces'] = '1'` as you've found. You can disable auto-pairing by adding to your *~/.textadept/init.lua*:

```
events.connect(events.LEXER_LOADED, function(lang)
  if lang == 'cpp' then textadept.editing.auto_pairs['<'] = nil end
end
```

(Basically, overriding https://github.com/orbitalquark/textadept/blob/81180915559448f60101f1d3da61819b40591e83/modules/textadept/editing.lua#L104-L110)

It would be nice to highlight `std::make_shared<int>()` as a function, but the fact that anything can be between `<...>` for templated things, lexing it is not straightforward.

As for treesitter, I experimented with Scintilla and treesitter (https://github.com/orbitalquark/tslexia), but nothing has come of it. A majority of compiled grammar shared objects are larger than 1MB in size *each*. Compare that to Scintillua lexers, which are between 1-5KB on average. All 100+ of them cumulatively total 1MB in size. Once I realized this, I kind of lost interest in trying to further explore this. Also, last I checked, developing and compiling treesitter grammars requires node.JS, which feels so dirty.

**Reply by snoopy** [reply]
Thanks, that works nicely. 
I was using `textadept.editing.auto_pairs = {}` which did not prevent auto pairing as well as I expected.

>developing and compiling treesitter grammars requires node.JS

I had no idea. That's really lame.
